referenceframe/worldstate.go

```diff
@@ -1,51 +1,62 @@
 package referenceframe
 
 import (
-	"fmt"
 	"strconv"
 
 	commonpb "go.viam.com/api/common/v1"
 
-	spatial "go.viam.com/rdk/spatialmath"
+	"go.viam.com/rdk/spatialmath"
 )
 
 // WorldState is a struct to store the data representation of the robot's environment.
 type WorldState struct {
-	Obstacles  []*GeometriesInFrame
-	Transforms []*LinkInFrame
+	names        map[string]bool
+	unnamedCount int
+	obstacles    []*GeometriesInFrame
+	Transforms   []*LinkInFrame
 }
 
 const unnamedWorldStateGeometryPrefix = "unnamedWorldStateGeometry_"
 
+func NewEmptyWorldState() *WorldState {
+	return &WorldState{
+		names:     make(map[string]bool),
+		obstacles: make([]*GeometriesInFrame, 0),
+	}
+}
+
 // WorldStateFromProtobuf takes the protobuf definition of a WorldState and converts it to a rdk defined WorldState.
 func WorldStateFromProtobuf(proto *commonpb.WorldState) (*WorldState, error) {
-	convertProtoGeometries := func(allProtoGeometries []*commonpb.GeometriesInFrame) ([]*GeometriesInFrame, error) {
-		list := make([]*GeometriesInFrame, 0, len(allProtoGeometries))
-		for _, protoGeometries := range allProtoGeometries {
-			geometries, err := ProtobufToGeometriesInFrame(protoGeometries)
-			if err != nil {
-				return nil, err
-			}
-			list = append(list, geometries)
-		}
-		return list, nil
-	}
 	if proto == nil {
-		return &WorldState{}, nil
-	}
-	obstacles, err := convertProtoGeometries(proto.GetObstacles())
-	if err != nil {
-		return nil, err
+		return NewEmptyWorldState(), nil
 	}
+
 	transforms, err := LinkInFramesFromTransformsProtobuf(proto.GetTransforms())
 	if err != nil {
 		return nil, err
 	}
 
-	return &WorldState{
-		Obstacles:  obstacles,
-		Transforms: transforms,
-	}, nil
+	ws := &WorldState{Transforms: transforms}
+	for _, protoGeometries := range proto.GetObstacles() {
+		geometries, err := ProtobufToGeometriesInFrame(protoGeometries)
+		if err != nil {
+			return nil, err
+		}
+		ws.AddObstacles(geometries.frame, geometries.geometries...)
+	}
+
+	return ws, nil
+}
+
+// AddObstacles takes in a list of geometries and a frame corresponding with the reference frame associated with them and adds them
+// as obstacles to the worldState
+func (ws *WorldState) AddObstacles(frame string, geometries ...spatialmath.Geometry) error {
```

Comment:
This should not be public, as we don't want it to be editable.

Reply:
I must be misunderstanding something.  We want to interact with the obstacles through this function.  It therefore should be public, no?

Reply:
We want to interact with the obstacles by retrieving them. We don't want to construct a worldstate piecemeal, and we don't want to add new obstacles to a worldstate after the fact. There should be no way for a user to alter a worldstate object that other things could be using.

Reply:
Would it fix things if this returned a new worldState instead of mutating it? 

I'm not sure why you say we don't want to construct it piecemeal.  Can you outline the way you are imagining this will look?

Reply:
I think the function signature should be
```
func NewWorldState(obstacles []*GeometriesInFrame, transforms []*LinkInFrame) *WorldState
```
rather than a method. Of course there can be a private method to add geometries (individually or all at once), but it shouldn't be able to be called from some random user-provided function halfway through motion planning the way this one is. As is it would affect anything holding onto a pointer to that worldstate.